content/en/blog/_posts/2021-12-15-pod-security-admission-psa-beta.md

```diff
@@ -0,0 +1,620 @@
+```
+
+Let's look at the default namespace.
+
+```shell
+kubectl describe namespace default
+```
+
+The output is similar to this:
+
+```shell
+Name:         default
+Labels:       kubernetes.io/metadata.name=default
+Annotations:  <none>
+Status:       Active
+
+No resource quota.
+
+No LimitRange resource.
+```
+
+Let's create a new namespace and see if the labels apply there.
+
+```shell
+kubectl create namespace test-defaults
+kubectl describe namespace test-defaults
+```
+
+Same.
+
+```shell
+Name:         test-defaults
+Labels:       kubernetes.io/metadata.name=test-defaults
+Annotations:  <none>
+Status:       Active
+
+No resource quota.
+
+No LimitRange resource.
+```
+
+Can we deploy a workload that is privileged?
+
+```shell
+cat <<EOF | kubectl -n test-defaults apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx-privileged
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+    securityContext:
+      allowPrivilegeEscalation: true
+EOF
+```
+
+Hmm... yep. Our default `warn` level is working at least.
+
+```shell
+Warning: would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
+pod/nginx-privileged created
+```
+
+Let's delete the pod with `kubectl -n test-defaults delete pod/nginx-privileged`.
+
+Is my config even working?
+
+```shell
+# shell into the control plane
+docker exec -it kind-control-plane bash
+
+# cat out the file we mounted
+cat /etc/kubernetes/policies/pod-security.yaml
+
+# check the api server logs
+cat /var/log/containers/kube-apiserver*.log 
+
+# check the api server config
+cat /etc/kubernetes/manifests/kube-apiserver.yaml 
+```
+
+**UPDATE:** The baseline policy permits `allowPrivilegeEscalation`. While I cannot see the Pod Security default levels of enforcement, they are there. Let's try to provide a manifest that violates the baseline by requesting hostNetwork access.
+
+```shell
+cat <<EOF | kubectl -n test-defaults apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx-privileged
+spec:
+  containers:
+  - name: nginx
+    image: nginx
+  hostNetwork: true
+EOF
+```
+
+The output is similar to this:
+
+```shell
+Error from server (Forbidden): error when creating "STDIN": pods "nginx-privileged" is forbidden: violates PodSecurity "baseline:latest": host namespaces (hostNetwork=true)
+```
+
+#### Yes!!! It worked! 🎉🎉🎉
+
+I later found out, another way to check if things are operating as intended is to check the raw API server metrics endpoint.
+
+Run the following command:
+
+```
+kubectl get --raw /metrics | grep pod_security_evaluations_total
+```
```

Comment:
Oh, right, nice. This is a better way to check if it's working/enabled (section above).

Reply:
Do you think it's worth stripping the user journey section of demo and consolidating it by starting with the `raw metrics` check to validate followed by a clean(er) working demo?

I like both approaches. I do enjoy watching "the journey" but other folks might want just raw / working content.

Reply:
Since this is a blog post, I like the user journey section (imo)